src/category_theory/idempotents/karoubi_karoubi.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2022 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+
+import category_theory.idempotents.karoubi
+
+/-!
+# Idempotence of the Karoubi envelope
+
+In this file, we construct the equivalence of categories
+`karoubi_karoubi.equivalence C : karoubi C ≌ karoubi (karoubi C)` for any category `C`.
+
+-/
+
+open category_theory.category
+open category_theory.idempotents.karoubi
+
+namespace category_theory
+
+namespace idempotents
+
+namespace karoubi_karoubi
+
+variables (C : Type*) [category C]
+
+/-- The canonical functor `karoubi (karoubi C) ⥤ karoubi C` -/
+@[simps]
+def inverse : karoubi (karoubi C) ⥤ karoubi C :=
+  { obj := λ P, ⟨P.X.X, P.p.f, by simpa only [hom_ext] using P.idempotence⟩,
+    map := λ P Q f, ⟨f.f.f, by simpa only [hom_ext] using f.comm⟩, }
+
+instance [preadditive C] : functor.additive (inverse C) := { }
+
+/-- The unit isomorphism of the equivalence -/
+@[simps]
+def unit_iso : 𝟭 (karoubi C) ≅ to_karoubi (karoubi C) ⋙ inverse C := eq_to_iso begin
+  apply functor.ext,
+  { intros P Q f,
+    ext,
+    simp only [functor.id_map, inverse_map_f, to_karoubi_map_f, eq_to_hom_f,
+      eq_to_hom_refl, comp_id, p_comp_assoc, functor.comp_map, comp],
+    dsimp,
+    simp only [id_eq, comp_p], },
+  { intro P,
+    ext,
+    { simpa only [eq_to_hom_refl, comp_id, id_comp], },
+    { refl, }, }
+end
+
+/-- The counit isomorphism of the equivalence -/
+@[simps]
+def counit_iso : inverse C ⋙ to_karoubi (karoubi C) ≅ 𝟭 (karoubi (karoubi C)) :=
+{ hom :=
+  { app := λ P,
+    { f :=
+      { f := P.p.1,
+        comm := begin
+          have h := P.idempotence,
+          simp only [hom_ext, comp] at h,
+          erw [← assoc, h, comp_p],
+        end, },
+      comm := begin
+        have h := P.idempotence,
+        simp only [hom_ext, comp] at h ⊢,
+        erw [h, h],
+      end, },
+    naturality' := λ P Q f, by simpa only [hom_ext] using (p_comm f).symm, },
+  inv :=
+  { app := λ P,
+    { f :=
+      { f := P.p.1,
+        comm := begin
+          have h := P.idempotence,
+          simp only [hom_ext, comp] at h,
+          erw [h, p_comp],
+        end, },
+      comm := begin
+        have h := P.idempotence,
+        simp only [hom_ext, comp] at h ⊢,
+        erw [h, h],
+      end, },
+    naturality' := λ P Q f, by simpa [hom_ext] using (p_comm f).symm, },
+  hom_inv_id' := by { ext P, simpa only [hom_ext, id_eq] using P.idempotence, },
+  inv_hom_id' := by { ext P, simpa only [hom_ext, id_eq] using P.idempotence, }, }
+
+/-- The equivalence `karoubi C ≌ karoubi (karoubi C)` -/
+@[simps]
+def equivalence : karoubi C ≌ karoubi (karoubi C) :=
```

Comment:
You already have
```
/-- If `C` is idempotent complete, the functor `to_karoubi : C ⥤ karoubi C` is an equivalence. -/
def to_karoubi_is_equivalence [is_idempotent_complete C] :
  is_equivalence (to_karoubi C) :=
equivalence.of_fully_faithfully_ess_surj (to_karoubi C)
```
But it has a non-explicit inverse. Is that the reason why you don't want to reuse it here?

Reply:
Yes, this is the reason. For the Dold-Kan equivalence, I will have to compose functors with this explicit inverse.

Reply:
Is there a short proof that `inverse` defined above is naturally isomorphic to the inverse rolling out of `is_equivalence`. Because then you might be able to prove `unit_iso` and `counit_iso` via transport of structure. Does that make sense?

Reply:
I am not 100% sure what you mean. What is clear to me is that as we already know that the functor `to_karoubi (karoubi C)` is an equivalence of categories, then in order to define `karoubi_karoubi.equivalence`, mathematically speaking, only the explicit inverse and the `unit_iso` (which is eq_to_iso) are the most relevant data. The counit_iso and the compatibility `functor_unit_iso_comp` should arise formally from the other information, but it does not seem that there is a general lemma in mathlib saying that if `is_equivalence F` and that if G is a functor in the other direction equipped with a `unit_iso 𝟭 C ≅ F ⋙ G`, then there is an equivalence with `inverse := G`.

Anyway, it this case, both unit_iso and counit_iso have quite straightforward definitions, and in case we need to evaluate the counit_iso of karoubi_karoubi.equivalence and dsimp, then it is probably better to have a down-to-earth definition rather than an indirect one.